### PR TITLE
Add top down keypoiont detection wrapper

### PR DESCRIPTION
**Pull request by sovrasov**
# What does this PR do?

A simple keypoint detection model wrapper is added. Only SimCC keypoint encoding is supported.

### CODE REVIEW

File: model_api/python/model_api/models/keypoint_detection.py [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+
+    Note:
+        instance number: N
+        num_keypoints: K
+        heatmap height: H
+        heatmap width: W
+
+    Args:
+        simcc_x (np.ndarray): x-axis SimCC in shape (K, Wx) or (N, K, Wx)
+        simcc_y (np.ndarray): y-axis SimCC in shape (K, Wy) or (N, K, Wy)
```
Comment (eunwoosh) [inline model_api/python/model_api/models/keypoint_detection.py:217]:
```suggestion
        simcc_y (np.ndarray): y-axis SimCC in shape (K, Hy) or (N, K, Hy)
```
I'm not sure but isn't it `H`?

File: model_api/python/model_api/models/keypoint_detection.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+            size. For example, if the input width is ``w``, the x label size
+            will be :math:`w*simcc_split_ratio`. Defaults to 2.0
+
+    .. _`SimCC: a Simple Coordinate Classification Perspective for Human Pose
+    Estimation`: https://arxiv.org/abs/2107.03332
+    """
+
+    def __init__(
+        self,
+        smoothing_type: str = "gaussian",
```
Comment (eunwoosh) [inline model_api/python/model_api/models/keypoint_detection.py:169]:
```suggestion
```
Is it necessary?

**Review by eunwoosh — COMMENTED** [review]
Thanks for your work. I left some comments. Please take a look.

File: model_api/python/model_api/models/keypoint_detection.py [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+                It usually represents the confidence of the keypoint prediction
+        """
+        keypoints, scores = get_simcc_maximum(simcc_x, simcc_y)
+
+        # Unsqueeze the instance dimension for single-instance results
+        if keypoints.ndim == 2:
+            keypoints = keypoints[None, :]
+            scores = scores[None, :]
+
+        keypoints /= self.simcc_split_ratio
```
[the remark this answers is not in the document]
Comment (sovrasov) [reply model_api/python/model_api/models/keypoint_detection.py:198]: / is right, the doc says the coordinate space is inflated by `simcc_split_ratio`. When decoding, we have to squeeze it back

File: model_api/python/docs/keypoint_detection.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+## How to use
+
+```python
+import cv2
+from model_api.models import TopDownKeypointDetectionPipeline, Detection, KeypointDetectionModel
+
+    model = KeypointDetectionModel.create_model("kp_model.xml")
+    detections = [Detection(0, 0, 100, 100, 0, 0)]
+    top_down_pipeline = TopDownKeypointDetectionPipeline(model)
+    predictions = top_down_detector.predict(image, det)
```
[the remark this answers is not in the document]
Comment (sovrasov) [reply model_api/python/docs/keypoint_detection.md]: Updated

File: model_api/python/model_api/models/keypoint_detection.py [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+
+
+class KeypointDetectionModel(ImageModel):
+    """
+    A wrapper that implements a basic keypoint regression model.
+    """
+
+    __model__ = "keypoint_detection"
+
+    def __init__(self, inference_adapter, configuration=dict(), preload=False):
```
[the remark this answers is not in the document]
Comment (sovrasov) [reply model_api/python/model_api/models/keypoint_detection.py:36]: Added

File: model_api/python/model_api/models/keypoint_detection.py [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+        self.simcc_split_ratio = simcc_split_ratio
+
+    def decode(
+        self, simcc_x: np.ndarray, simcc_y: np.ndarray
+    ) -> tuple[np.ndarray, np.ndarray]:
+        """Decode keypoint coordinates from SimCC representations. The decoded coordinates are in the input image space.
+
+        Args:
+            encoded (Tuple[np.ndarray, np.ndarray]): SimCC labels for x-axis
+                and y-axis
```
[the remark this answers is not in the document]
Comment (sovrasov) [reply model_api/python/model_api/models/keypoint_detection.py]: Implementation refactored

File: model_api/python/model_api/models/keypoint_detection.py [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+
+        - keypoint_x_labels (np.ndarray): The generated SimCC label for x-axis.
+            The label shape is (N, K, Wx) if ``smoothing_type=='gaussian'``
+            and (N, K) if `smoothing_type=='standard'``, where
+            :math:`Wx=w*simcc_split_ratio`
+        - keypoint_y_labels (np.ndarray): The generated SimCC label for y-axis.
+            The label shape is (N, K, Wy) if ``smoothing_type=='gaussian'``
+            and (N, K) if `smoothing_type=='standard'``, where
+            :math:`Wy=h*simcc_split_ratio`
+        - keypoint_weights (np.ndarray): The target weights in shape (N, K)
```
[the remark this answers is not in the document]
Comment (sovrasov) [reply model_api/python/model_api/models/keypoint_detection.py]: Implementation refactored

File: model_api/python/model_api/models/keypoint_detection.py (hunk 1 above)
Comment (sovrasov) [reply model_api/python/model_api/models/keypoint_detection.py]: Right

File: model_api/python/model_api/models/keypoint_detection.py [hunk 8]
```diff
@@ -34,9 +34,17 @@ class KeypointDetectionModel(ImageModel):
     __model__ = "keypoint_detection"
 
     def __init__(self, inference_adapter, configuration=dict(), preload=False):
+        """
+        Initializes the keypoint detection model.
+
+        Args:
+            inference_adapter (InferenceAdapter): inference adapter containing the underlying model.
```
Comment (eunwoosh) [inline model_api/python/model_api/models/keypoint_detection.py:41]:
```suggestion
            inference_adapter (InferenceAdapter): Inference adapter containing the underlying model.
```

**Review by eunwoosh — COMMENTED** [review]
LGTM. I left one minor comment. Please take a look :)

**Review by eunwoosh — APPROVED** [review]

**Pull request merged**